### ISSUE TITLE
ERROR: Error during SonarScanner execution java.lang.IllegalStateException: Cannot register highlighting rule for characters when scanning directory

### ISSUE THREAD

**Issue by CartoonFan** [SonarOpenCommunity/sonar-cxx#2058]
Is it really a **cxx-plugin** issue or a general SonarQube issue? To track SonarQube issues use the [community from SonarSource](https://community.sonarsource.com/) please.

Please [turn debug info on](https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Get-debug-information) and reproduce your issue.
The debug info in the log file could help to solve or locate the issue.

### Description

I get an error: "ERROR: Error during SonarScanner execution
java.lang.IllegalStateException: Cannot register highlighting rule for characters at Range[from [line=25, lineOffset=17] to [line=25, lineOffset=21]] as it overlaps at least one existing rule" when trying to scan the master branch of [aosdk](https://github.com/nmlgc/aosdk) using the most recent master git checkout (https://github.com/SonarOpenCommunity/sonar-cxx/commit/b4fe02562332d71352272afbcb8660078ac120e9) of sonar-cxx.

### Steps to reproduce the problem

Please provide the steps required to reproduce the problem

1. Set sonar.sources to .

2. Scan the directory

Unfortunately, I'm not able to isolate which file causes the problem (likely a header, but I'm not sure) because scanning the file in question does not produce any errors.

If possible then please create a git repository with a repro sample or attach a zip to the issue.

### Expected behavior
A successful scan.

### Actual behavior

The scan will proceed normally with lots of "ERROR: Unable to parse file:" and " Lexer error: Unable to lex url:" outputs, until hitting the highlighting error right at the end.

### Known workarounds

Haven't found any yet, except excluding the file(s) in question from the scan.

### LOG file

Yes. Besides the above errors, there are also multiple "Highlighting error" outputs. The scanner also complains about a nonexistent include directory, but that's because I have that setup as a default include directory.  I have the log right here:
[sonar-scanner.log](https://github.com/SonarOpenCommunity/sonar-cxx/files/5968494/sonar-scanner.log)

### Related information

* cxx plugin version? https://github.com/SonarOpenCommunity/sonar-cxx/commit/b4fe02562332d71352272afbcb8660078ac120e9
* SonarQube version? Community Edition - Version 8.6.1 (build 40680)

**Comment by guwirth** [comment]
Hi @CartoonFan,

thanks for the feedback.

The version you are using is: **C++ (Community) 2.0.0-SNAPSHOT (cxx)**

There are some syntax errors. Most likely missing macros. Can you provide sample code?
```
12:58:07.772 DEBUG: [/home/jeremiah/workshop/aosdk/zlib/deflate.c:204]:    syntax error: int deflateInit_ ( strm , level , version , stream_size ) 
12:58:08.741 DEBUG: [/home/jeremiah/workshop/aosdk/zlib/inffast.c:67]:    syntax error: void inflate_fast ( strm , start ) 
```
Issue 1: Should not throw an exception if dir does not exist
```
java.nio.file.NoSuchFileException: /home/jeremiah/workshop/aosdk/include
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixPath.toRealPath(UnixPath.java:852)
	at org.sonar.cxx.preprocessor.SourceCodeProvider.setIncludeRoots(SourceCodeProvider.java:67)
	at org.sonar.cxx.preprocessor.CxxPreprocessor.addGlobalIncludeDirectories(CxxPreprocessor.java:649)
	at org.sonar.cxx.preprocessor.CxxPreprocessor.init(CxxPreprocessor.java:185)
	at com.sonar.sslr.impl.Lexer.initPreprocessors(Lexer.java:179)
	at com.sonar.sslr.impl.Lexer.lex(Lexer.java:119)
	at com.sonar.sslr.impl.Lexer.lex(Lexer.java:92)
	at com.sonar.sslr.impl.Lexer.lex(Lexer.java:81)
	at com.sonar.sslr.impl.Parser.parse(Parser.java:65)
	at org.sonar.squidbridge.AstScanner.scanFiles(AstScanner.java:94)
	at org.sonar.plugins.cxx.CxxSquidSensor.execute(CxxSquidSensor.java:277)
```

Issue 2: Maybe inherited error, but should also not throw
```
java.lang.IllegalStateException: Cannot register highlighting rule for characters at Range[from [line=25, lineOffset=17] to [line=25, lineOffset=21]] as it overlaps at least one existing rule
	at org.sonar.api.batch.sensor.highlighting.internal.DefaultHighlighting.checkOverlappingBoundaries(DefaultHighlighting.java:60)
	at org.sonar.api.batch.sensor.highlighting.internal.DefaultHighlighting.doSave(DefaultHighlighting.java:120)
	at org.sonar.api.batch.sensor.internal.DefaultStorable.save(DefaultStorable.java:45)
	at org.sonar.plugins.cxx.CxxSquidSensor.saveHighlighting(CxxSquidSensor.java:477)
	at org.sonar.plugins.cxx.CxxSquidSensor.save(CxxSquidSensor.java:349)
	at org.sonar.plugins.cxx.CxxSquidSensor.execute(CxxSquidSensor.java:280)
```

Regards,

**Comment by CartoonFan** [comment]
Hello, @guwirth! Thanks for responding. I'm still new at this, but it appears to be a version of zlib bundled with the program. I'm not sure exactly what you mean by sample code, but I'll try posting some code that may hopefully be relevant.

deflate.c (203:313):
```
/* ========================================================================= */
int ZEXPORT deflateInit_(strm, level, version, stream_size)
    z_streamp strm;
    int level;
    const char *version;
    int stream_size;
{
    return deflateInit2_(strm, level, Z_DEFLATED, MAX_WBITS, DEF_MEM_LEVEL,
                         Z_DEFAULT_STRATEGY, version, stream_size);
    /* To do: ignore strm->next_in if we use it as window */
}

/* ========================================================================= */
int ZEXPORT deflateInit2_(strm, level, method, windowBits, memLevel, strategy,
                  version, stream_size)
    z_streamp strm;
    int  level;
    int  method;
    int  windowBits;
    int  memLevel;
    int  strategy;
    const char *version;
    int stream_size;
{
    deflate_state *s;
    int wrap = 1;
    static const char my_version[] = ZLIB_VERSION;

    ushf *overlay;
    /* We overlay pending_buf and d_buf+l_buf. This works since the average
     * output size for (length,distance) codes is <= 24 bits.
     */

    if (version == Z_NULL || version[0] != my_version[0] ||
        stream_size != sizeof(z_stream)) {
        return Z_VERSION_ERROR;
    }
    if (strm == Z_NULL) return Z_STREAM_ERROR;

    strm->msg = Z_NULL;
    if (strm->zalloc == (alloc_func)0) {
        strm->zalloc = zcalloc;
        strm->opaque = (voidpf)0;
    }
    if (strm->zfree == (free_func)0) strm->zfree = zcfree;

#ifdef FASTEST
    if (level != 0) level = 1;
#else
    if (level == Z_DEFAULT_COMPRESSION) level = 6;
#endif

    if (windowBits < 0) { /* suppress zlib wrapper */
        wrap = 0;
        windowBits = -windowBits;
    }
#ifdef GZIP
    else if (windowBits > 15) {
        wrap = 2;       /* write gzip wrapper instead */
        windowBits -= 16;
    }
#endif
    if (memLevel < 1 || memLevel > MAX_MEM_LEVEL || method != Z_DEFLATED ||
        windowBits < 8 || windowBits > 15 || level < 0 || level > 9 ||
        strategy < 0 || strategy > Z_RLE) {
        return Z_STREAM_ERROR;
    }
    if (windowBits == 8) windowBits = 9;  /* until 256-byte window bug fixed */
    s = (deflate_state *) ZALLOC(strm, 1, sizeof(deflate_state));
    if (s == Z_NULL) return Z_MEM_ERROR;
    strm->state = (struct internal_state FAR *)s;
    s->strm = strm;

    s->wrap = wrap;
    s->w_bits = windowBits;
    s->w_size = 1 << s->w_bits;
    s->w_mask = s->w_size - 1;

    s->hash_bits = memLevel + 7;
    s->hash_size = 1 << s->hash_bits;
    s->hash_mask = s->hash_size - 1;
    s->hash_shift =  ((s->hash_bits+MIN_MATCH-1)/MIN_MATCH);

    s->window = (Bytef *) ZALLOC(strm, s->w_size, 2*sizeof(Byte));
    s->prev   = (Posf *)  ZALLOC(strm, s->w_size, sizeof(Pos));
    s->head   = (Posf *)  ZALLOC(strm, s->hash_size, sizeof(Pos));

    s->lit_bufsize = 1 << (memLevel + 6); /* 16K elements by default */

    overlay = (ushf *) ZALLOC(strm, s->lit_bufsize, sizeof(ush)+2);
    s->pending_buf = (uchf *) overlay;
    s->pending_buf_size = (ulg)s->lit_bufsize * (sizeof(ush)+2L);

    if (s->window == Z_NULL || s->prev == Z_NULL || s->head == Z_NULL ||
        s->pending_buf == Z_NULL) {
        s->status = FINISH_STATE;
        strm->msg = (char*)ERR_MSG(Z_MEM_ERROR);
        deflateEnd (strm);
        return Z_MEM_ERROR;
    }
    s->d_buf = overlay + s->lit_bufsize/sizeof(ush);
    s->l_buf = s->pending_buf + (1+sizeof(ush))*s->lit_bufsize;

    s->level = level;
    s->strategy = strategy;
    s->method = (Byte)method;

    return deflateReset(strm);
}

/* ========================================================================= */
```
deflate.h(90:91):
```
typedef struct internal_state {
    z_streamp strm;      /* pointer back to this zlib stream */
```
deflate.h (177):
`int level;    /* compression level (1..9) */`

I wasn't able to find version or stream_size in a header file that was directly included (the only one that might have it is zlib.h).

inffast.c (67):
```
void inflate_fast(strm, start)
z_streamp strm;
unsigned start;         /* inflate()'s starting value for strm->avail_out */
```
I'm by no means an expert, but looking at the functions in question, it seems that it's written like a weird hybrid between a function call and a function definition. Maybe that's part of the issue? Honestly, I'm not sure. I also wanted to add that Issue 2 happens even on a branch that has an include directory present. Hopefully this information helps! I'll try to provide any additional information if possible.

**Comment by guwirth** [comment]
Hello @CartoonFan ,

the reason for the syntax error below is the missing `ZEXPORT` macro:
```C
int ZEXPORT deflateInit_(strm, level, version, stream_size)
```
You have to ensure that the header file with the definition can be loaded or even better "remove" it with a define:
```C
#define ZEXPORT // define it e.g in sonar.cxx.defines
int ZEXPORT deflateInit_(strm, level, version, stream_size)
```
Please read also 
https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Detect-and-fix-parsing-errors
https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Dealing-with-compiler-specific-code-pieces

Regards,

**Comment by CartoonFan** [comment]
Thanks, @guwirth! Adding a define for ZEXPORT (which seems to be empty) fixes the issue and the scan completes successfully 🥳!

Closing #2058!

**Comment by guwirth** [comment]
Like to keep this open to have a look to the extensions.

**Comment by CartoonFan** [comment]
OK, that's fine.
